primitives/src/crh/poseidon/batched_crh.rs

```diff
@@ -0,0 +1,537 @@
+            let mut idx: i64 = w.len() as i64 - P::R as i64;
+            for i in (0..vec_state.len()).rev() {
+                let vec_1 = vec_state[i][0].clone();
+                vec_state[i][0] = w_bar * &w[idx as usize];
+                w_bar = w_bar * &vec_1;
+                idx -= 1;
+            }
+        }
+    }
+
+    pub fn poseidon_perm_gen(vec_state: &mut [Vec<P::Fr>]) {
+
+        // index that goes over the round constants
+        let mut round_cst_idx: usize = 0;
+
+        // Full rounds
+        for _i in 0..P::R_F {
+            Self::poseidon_full_round(vec_state, &mut round_cst_idx);
+
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+
+        }
+
+        // Partial rounds
+        for _i in 0..P::R_P {
+            Self::poseidon_partial_round(vec_state, &mut round_cst_idx);
+
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+        }
+
+        // Full rounds
+        // Last round does not contain the matrix mix
+        for _i in 0..P::R_F {
+            Self::poseidon_full_round(vec_state, &mut round_cst_idx);
+
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+        }
+    }
+}
+
+
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr = F>> BatchFieldBasedHash for PoseidonBatchHash<F, P> {
+    type Data = F;
+    type BaseHash = PoseidonHash<F, P>;
+
+    fn batch_evaluate(input_array: &[F]) -> Result<Vec<F>, Error> {
+
+        // Input:
+        // This function calculates the hashes of inputs by groups of the rate P::R.
+        // The inputs are arranged in an array and arranged as consecutive chunks
+        // Example:
+        // For P::R = 2,
+        // (d_00, d_01, d_10, d_11, d_20, d_21, ...
+        // where d_00 and d_01 are the inputs to the first hash
+        // d_10 and d_11 are the inputs to the second hash ... etc..
+        // Output:
+        // The output is returned as an array of size input length / P::R
+
+        use rayon::prelude::*;
+
+        // Checks that size of input/output vector
+        assert_eq!(P::T - P::R, 1, "The assumption that the capacity is one field element is not satisfied.");
+        let array_length = input_array.len() / P::R;
+        assert_eq!(input_array.len() % P::R, 0, "The length of the input data array is not a multiple of the rate.");
+        assert_ne!(input_array.len(), 0, "Input data array does not contain any data.");
+
+        // Assign pre-computed values of the state vector equivalent to a permutation with zero element state vector
+        let mut state_z = Vec::new();
+        for i in 0..P::T {
+            state_z.push(P::AFTER_ZERO_PERM[i]);
+        }
```

Comment:
I can't find this procedure in the current version of the paper. I have to double-check, but I think we can remove it (and the AFTER_ZERO_PERM from the trait).

Reply:
I think there was a good reason for doing it, please before removing it let's double check

Reply:
The only reason I figured out is the cpp implementation of a full Poseidon sponge in an [old commit](https://extgit.iaik.tugraz.at/krypto/hadeshash/-/tree/8cbd9eb13c06ae48b7038f304e6c43a04f9c90fc)  of the Hadeshash repository. 
I double checked the new paper: In Appendix I they explicitly describe the node hash function of a Poseidon based Merkle tree. Besides that it does a different treatment for handling empty (or, "phantom") leaves, it clearly does not do this extra permutation. 
Although keeping it is no security issue, I'd opt for removing it.